### DISCUSSION TITLE
MP 1 Q&A

### DISCUSSION THREAD

**Discussion by Abdeltif1** [SOEN345-WINTER2024/discussion#7] [Q&A]
Hello everyone!

This is going to be our discussion form. You can ask your questions here and we will discuss and answer them. 

Happy coding!

**Comment by inas-fawzi** [comment]
The github repo says the assignment is due feb 16 but the moodle says it is due feb 17. When is it due?

**Reply by Abdeltif1** [reply]
Hello Inas, 
Thank you for reaching out, the assignment is due  Feb 17, 2024, 11:49 EST.

**Reply by inas-fawzi** [reply]
Thank you, also is it normal to be getting these errors right after cloning the repo?
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/95307617/c4addf95-da91-4607-b821-63dbf034f9c8)

**Reply by Abdeltif1** [reply]
I will be making an announcement soon about how to properly build the project and what to do to avoid such errors

**Reply by inas-fawzi** [reply]
okay thanks

**Comment by Abdeltif1** [comment]
How to properly build your BasicCalculator:

1. After you clone and load the project on android studio, go to **File -> Settings -> Build, Execution, Deployment -> Build Tools -> Gradle**
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/45ee0b92-df71-4e63-8bab-c42a918d818c)
2. Then click on **Gradle JDK  -> Download JDK -> Download JDK 11** 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/6884a99b-06d8-4806-88ff-da3aade5fdd1)
3. Once it is downloaded make sure it is selected: 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/76917d2f-6b7a-4439-bfdc-7701d02a4f02)
4. After that go to **File -> Project Structure**  and change Gradle version to 8.6: 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/18e84cf4-235b-41fc-9b0b-4029353d7224)
5. Then click okay and go to: **Gradle Scripts -> build.gradle** and change it to this (from 3.x.x to 7.4.1) type it
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/085fa357-58d7-4c69-a7ac-081bb44764e0)
6.Then look at the bottom of your screen, go to Build -> Sync and click on the sync button: 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/e8372076-8aca-4502-b5fa-ff59242649a6)
7. Then you click on Build output and the build button and the project should be built successfully! 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/5ad05dc2-e220-4896-81b2-e2d4e68e39b1)

You then should be able to run the application: 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/bd64ad03-faab-4fd6-a549-88e1ec36b0d1)



And you should be good to go!

**Reply by amirhossein942** [reply]
After cloning the project, I don't even see the project under the gradle projects in the first photo, I followed the instruction's PDF but nothing works. It just runs and says execution completed without displaying the calculator.

**Reply by Abdeltif1** [reply]
Hello @amirhossein942 we will assist you today during the lab to have it built properly.

**Reply by Abdeltif1** [reply]
@amirhossein942 are you sure this is your section? are you with Dr. Ali?

**Reply by amirhossein942** [reply]
no actually I'm in the wrong section

**Reply by Abdeltif1** [reply]
@amirhossein942 that is fine, please re-clone the repository and follow the steps above, it should work for you. if you need any further help, I'd be happy to assist you.

**Reply by Facu-alfaro** [reply]
Hi, is there a different way to do it on macOS? Because I can't find most of the things that you showed.

**Reply by Abdeltif1** [reply]
@Facu-alfaro hey, I tried the same thing with other students on Mac today and it worked just fine, make sure that when you open the project on android studio you are opening BasicApp not mp1-.... for it to work fine.

**Reply by zaheerqur** [reply]
This helped, thankyou

**Comment by im-samwong** [comment]
https://classroom.github.com/classrooms/151584622-soen345-winter24-classroom-c3febe

This link is returning 404.

**Reply by Abdeltif1** [reply]
Hello, can you tell me what are you trying to access and I will give you the right link, if you want access to the assignment MP1, here you go: https://classroom.github.com/a/lyf7DF2M

**Comment by CritixGames** [comment]
I am at the last step of running my virtual pixel phone but I get this error and it is really annoying me that it is this complicated (It wants me to install the Running Android Emulator hypervisor driver) : 

Android SDK is up to date.
Running Android Emulator hypervisor driver installer
[SC] ControlService FAILED 1062:

The service has not been started.

[SC] DeleteService SUCCESS
[SC] StartService FAILED with error 4294967201.
Done
 
thank you.

**Reply by Abdeltif1** [reply]
Hello @CritixGames, did the project build successfully? Also are you using AMD Ryzen? 
We will assist you today during the lab.

**Reply by CritixGames** [reply]
I am on my desktop computer can we setup a zoom?

**Reply by Abdeltif1** [reply]
@CritixGames, please share you issue with me and I can assist you here in case it helps everyone.

**Comment by Yason287** [comment]
![Screenshot (915)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/c9d76b9d-e5e0-467b-aef6-7bc46a899437)
Hello, I have a problem. Could you help me?

**Reply by Abdeltif1** [reply]
Hello @Yason287, can you sync, click on the spiral button on the left under sync and it should work if you followed the steps above

**Reply by Yason287** [reply]
![Screenshot (916)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/583a24ee-d8e8-4bb7-9f12-2c207c0df5ef)
I got the same result. It seems  followed all the given steps. I have also noticed that near Build and Sync, I don't have Build Out.

**Reply by Abdeltif1** [reply]
if you can see at the top of your screen, there's a build button click on it then make project

**Reply by Yason287** [reply]
![Screenshot (917)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/537346fa-0c95-4b16-8c67-9b2071bf0f4a)
I tried, but it didn't work with compileSdrversion 27. Then I downloaded 30.0.3 and I got the result above.

**Reply by Abdeltif1** [reply]
@Yason287 can you set it to 27 again, and just try to run the code after it build, by run I mean press the start button to emulate it, if the issue still occurs, I can surely help you on Monday in person or via zoom if you can email me.

**Reply by Yason287** [reply]
![Screenshot (918)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/c0684dbe-751c-4e22-918e-965323da98c3)

**Reply by Abdeltif1** [reply]
Please send me an email, abdeltif.azzizi@concordia.ca and we can sit together and fix it.

**Reply by Yason287** [reply]
Thanks, I just sent an email.

**Comment by Yason287** [comment]
![Screenshot (920)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/2e78405a-15c1-4e0c-9d55-b8f8369822f1)

I figured out a problem, I just needed to download Android 8.1 ("Oreo"). 
Is the problem above because my laptop is too old? I just have 8 GB RAM.

**Reply by Yason287** [reply]
![Screenshot (921)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/a8de57e1-8cd2-4bc9-a510-ba9bb9a64bd3)

**Comment by hisham-kitaz** [comment]
Hey, AbdelLatif  I am not super familiar with writing TDD do you think this youtube video shows it accurately https://www.youtube.com/watch?v=zFJdQYn9u_8

**Reply by Abdeltif1** [reply]
@hisham-kitaz Hello Hisham!, in short TDD is a software development practice that focuses on creating unit test cases before developing the actual code. It is an iterative approach combining programming, unit test creation, and refactoring. So in the assignment you will have to create tests, commit then write the code that you wrote the test cases for, and yes the video is a good resource. I hope this helps! 
if you have any other questions, let me know.

**Reply by hisham-kitaz** [reply]
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/111257875/3fa5ef7f-a2a7-4564-8820-17e480708140)
I don't understand what these two points mean. for point 5 I don't get what needs to be done. For point 6 after 5 is done how are we supposed to do the testing on what method are we testing since the only method we have is "addDecimal" which takes a string, then how do we test on the arithmetic operations? I'm confused.

**Reply by Abdeltif1** [reply]
Hello hisham, test cases are not written for you, you have to write them before you implement your functions this is TDD, we usually test methods, same as in the lab, so after you write your test case for a method that takes decimals, you then implement it in your MainActivity and test it against the test cases you wrote. In short, you have to write functions after you write your test cases. I hope this helps. I will explain it further during lab time.

**Comment by ibrahimdaami** [comment]
Hello, for question 6  Add one JUnit test for each arithmetic operation and commit this test. Are we supposed to add the tests then make them pass or just commit the tests?
Also for the submission it says to not submit apk files do we just add *.apk to gitignore and that should do it?

**Reply by ibrahimdaami** [reply]
Also what's the difference between 5 and 7

**Reply by Abdeltif1** [reply]
Hello @mfalmemkuu, yes you have to make them pass, that is the point of writing them. Once you are done, you can push, I believe git will handle all of that for you, I configured it to take only the necessary files for us to run your code. 

Good luck.

**Comment by Yason287** [comment]
![Screenshot (1)](https://github.com/SOEN345-WINTER2024/discussion/assets/78317885/779772cf-4f7b-4bef-b353-f79fa9aa9437)

I have a question about location of the files committed and messages. Is the image above is correct for question 3? 
1)Should I name the file and message to commit  "Add TDD test1"?
2) For the /* Initial test for TDD */, it should be inside of the code, right?

**Reply by Abdeltif1** [reply]
if you click on history you can see the following, all your versions: 
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/73204352/3186a8dc-5d0e-4302-be9c-027b8744cbc3)

**Reply by Yason287** [reply]
Is it correct?

**Reply by Yason287** [reply]
Could you answer on my questions? Did I properly commit for question 3? 
I still have two these questions:
1)Should I name the file and message to commit "Add TDD test1"?
2) For the /* Initial test for TDD */, it should be inside of the code, right?

**Comment by PaulYH** [comment]
Hello, does anything need to be submitted on Moodle, or is simply pushing all the work to our GitHub repo enough?

I just wanted to have this clarified because the assignment's Moodle page is accepting submissions:
![image](https://github.com/SOEN345-WINTER2024/discussion/assets/98189487/a62c16f5-7907-447f-b3b3-a83b0c73e111)
